### ISSUE TITLE
Can we fix toLower/toUpper?

### ISSUE THREAD

**Issue by pbiggar** [darklang/dark#3264]
The .NET toUpper/toLower do not handle some unicode issues as well as the OCaml code. I believe .NET handles "simple" Case Mapping, but not "full" Case Mapping.

[This comment](https://github.com/darklang/dark/blob/3c76b0b7375e0ca2fb47df76b96b33fcb6ada36e/fsharp-backend/tests/testfiles/string.tests#L518) describes what I know.

It appears that someone wrote some code that might solve this: https://github.com/dotnet/runtime/issues/30960#issuecomment-847502414 (links to https://gist.github.com/tarekgh/de1a1b29b03d048ad580c0a338c44fbb). We should determine if this solves it, and if so, incorporate it in `String::toUpper` and `String::toLower`

**Comment by StachuDotNet** [comment]
> We should determine if this solved it

To check this, I:
- created a new .sln locally, with C# src and F# test projects in it
- copied the C# code from the gist above into the src project
- wrote a few lines of F# to test the exposed fn there, against the relevant OCAMLONLY test cases [here](https://github.com/darklang/dark/blob/3c76b0b7375e0ca2fb47df76b96b33fcb6ada36e/fsharp-backend/tests/testfiles/string.tests#L534-L537)

And seemingly this doesn't not solve our situation? Or perhaps I am misinterpreting.

**Comment by pbiggar** [comment]
I am quite surprised it didn't work - the comments seemed to indicate a high degree of confidence that it would. But you are indeed testing the right thing afaict.

**Comment by StachuDotNet** [comment]
I was surprised too.
For what it's worth, I created a separate project in case you'd like to clone: https://github.com/StachuDotNet/StringCasingFix
![image](https://user-images.githubusercontent.com/906686/159257952-4f01a405-5d5d-48a4-a9f1-84a538c9c9eb.png)

Maybe it's worth mentioning in the other thread? Maybe it's specific to .NET 6?

**Comment by pbiggar** [comment]
> Maybe it's worth mentioning in the other thread? Maybe it's specific to .NET 6?

For sure!

**Comment by StachuDotNet** [comment]
> > Maybe it's worth mentioning in the other thread? Maybe it's specific to .NET 6?
> 
> For sure!

I have a draft and would appreciate some feedback on the below, @pbiggar - still finding myself a bit lost here!

---

I'm attempting to use [the provided gist](https://gist.github.com/tarekgh/de1a1b29b03d048ad580c0a338c44fbb) to resolve the same (or seemingly such) issue of surprising `ToUpper` results noted [here](https://github.com/dotnet/runtime/issues/30960#issuecomment-535274401), in darklang/dark#3264, linked just before this comment.

Namely, we're hoping to find a solution that will `ToUpper` strings like `"և"` and `"ﬁﬂ"` properly (to `"ԵՒ"` and `"FIFL"`, respectively).

I've copied the code into a new .net6 project, only to find that they're not working the same as noted in 
https://github.com/StachuDotNet/StringCasingFix

This is a bit tangential, so hope I'm not distracting much, but was wondering if I'm missing something - should I expect this to work? Has something changed, perhaps related to .NET versions?

**Comment by StachuDotNet** [comment]
Hm. Their example of `ﬃ`, in my solution, yields `'ﬃ` (no-op). That feels like the gist simply doesn't work in [my OS + .NET version + ???]. So the gist and _that comment_ aren't in sync, and I/we were mistakenly assuming they were.

**Comment by StachuDotNet** [comment]
FWIW, on .net core 3.1, I get the same results:
```
Expected: 'TEST'; Actual: 'TEST'
Expected: 'FIFL'; Actual: 'ﬁﬂ'
Expected: 'FFL'; Actual: 'ﬃ'
Expected: 'ԵՒ'; Actual: 'և'
```

I could set up .NET on my Windows machine and try there to rule out OS? Seems somewhat overkill, though, and now just thinking the gist doesn't match what we're looking for.

**Comment by pbiggar** [comment]
Yeah. Maybe let them know and we can move on if they don't respond.

**Comment by pbiggar** [comment]
Whoops!

What should we do here then? Close this issue?

**Comment by StachuDotNet** [comment]
I've scoured the internet the best I can to find another solution here, but haven't found anything.
Theoretically, this is still possible, I presume, so closing feels a bit wrong, but likely not blocking for an F# release. So maybe let's leave the conversation here and leave it for "some day"?

**Comment by StachuDotNet** [comment]
Should this be closed in favor of #3437? (or the other way around)
